platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java

```diff
@@ -211,11 +211,15 @@ public void variants(Action<? super NamedDomainObjectContainer<ConfigurationVari
     @Override
     public void capability(Object notation) {
         if (canCreate) {
-            Capability descriptor = capabilityNotationParser.parseNotation(notation);
             if (capabilities == null) {
-                capabilities = Lists.newArrayListWithExpectedSize(1); // it's rare that a component would declare more than 1 capability
+                capabilities = domainObjectCollectionFactory.newDomainObjectCollection(Capability.class);
+            }
+            if (notation instanceof Provider) {
```

Comment:
What are your thoughts on somehow moving this logic into the CapabilityNotationParserFactory? Right now, if I attempt to parse an incompatible type (for example pass an `Object` into this method), I get the following error:

```
Cannot convert the provided notation to an object of type Capability: java.lang.Object@6ccac6f4.
The following types/formats are supported:
  - Instances of Capability.
  - Instances of CharSequence.
  - Maps, for example [group: 'org.group', name: 'capability', version: '1.0'].
```

This message does not advertise that we support providers.

Reply:
I think we just have to live with that for now, perhaps we could add some flag in `NotationParser` to indicate that providers are mapped using it at a higher level, or we'd have to make the `NotationParser` produce a `Provider`, which would be more involved and possibly have issues with generics.

Reply:
If we don't add native support to the notation parser, we will also likely need to update the other places the parser is used.

Most importantly:
[DefaultMutableModuleDependencyCapabilitiesHandler](https://github.com/gradle/gradle/blob/1b03de5223aaea17cf044b88aa64af3c21a43d9b/subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultMutableModuleDependencyCapabilitiesHandler.java#L35-L37)

Maybe:
[DefaultCapabilitiesResolution](https://github.com/gradle/gradle/blob/1b03de5223aaea17cf044b88aa64af3c21a43d9b/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolutionstrategy/DefaultCapabilitiesResolution.java#L155)

Reply:
Done.